File: yasmin_factory/src/yasmin_factory_node.cpp
// Copyright (C) 2025 Miguel Ángel González Santamarta
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.

#include <iostream>
#include <memory>
#include <string>

#include "ament_index_cpp/get_package_share_directory.hpp"
#include "rclcpp/rclcpp.hpp"
#include "yasmin/state_machine.hpp"
#include "yasmin_factory/yasmin_factory.hpp"
#include "yasmin_ros/ros_logs.hpp"
#include "yasmin_ros/yasmin_node.hpp"
#include "yasmin_viewer/yasmin_viewer_pub.hpp"

int main(int argc, char *argv[]) {
  YASMIN_LOG_INFO("yasmin_factory_node");
  rclcpp::init(argc, argv);

  // Set up ROS 2 loggers
  yasmin_ros::set_ros_loggers();

  // Get the state machine file parameter
  auto node = yasmin_ros::YasminNode::get_instance();
  node->declare_parameter("state_machine_file", "");
  std::string sm_file = node->get_parameter("state_machine_file")
                            .get_parameter_value()
                            .get<std::string>();

  // Create the factory in a scope
  yasmin_factory::YasminFactory factory;

  // Create the state machine from the XML file
  auto sm = factory.create_sm_from_file(sm_file);

  // Cancel state machine on ROS 2 shutdown
  std::weak_ptr<yasmin::StateMachine> weak_sm = sm;
  rclcpp::on_shutdown([weak_sm]() {
    if (auto sm = weak_sm.lock()) {
      if (sm->is_running()) {
        sm->cancel_state();
      }
    }
  });

  // Publisher for visualizing the state machine
  yasmin_viewer::YasminViewerPub yasmin_pub(sm);

  // Execute the state machine
  try {
    std::string outcome = (*sm.get())();
    YASMIN_LOG_INFO(outcome.c_str());
  } catch (const std::exception &e) {
    YASMIN_LOG_WARN(e.what());
  }

  // Shutdown ROS 2
  rclcpp::shutdown();
  return 0;
}

File: yasmin_factory/yasmin_factory/yasmin_factory_node.py
#!/usr/bin/env python3

# Copyright (C) 2025 Miguel Ángel González Santamarta
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import rclpy
from yasmin_ros.yasmin_node import YasminNode

import yasmin
from yasmin_factory import YasminFactory
from yasmin_ros import set_ros_loggers
from yasmin_viewer import YasminViewerPub


def main() -> None:
    # Initialize ROS 2
    rclpy.init()

    # Get the state machine file parameter
    node = YasminNode.get_instance()
    node.declare_parameter("state_machine_file", "")
    sm_file = node.get_parameter("state_machine_file").get_parameter_value().string_value

    # Get if enable viewer parameter
    node.declare_parameter("enable_viewer_pub", True)
    enable_viewer_pub = (
        node.get_parameter("enable_viewer_pub").get_parameter_value().bool_value
    )

    # Set ROS 2 loggers
    set_ros_loggers()
    yasmin.YASMIN_LOG_INFO("yasmin_factory_node")

    # Create a finite state machine (FSM)
    factory = YasminFactory()
    sm = factory.create_sm_from_file(sm_file)
    sm.set_sigint_handler(True)

    # Publish FSM information for visualization
    pub = None
    if enable_viewer_pub:
        pub = YasminViewerPub(sm)

    # Execute the FSM
    try:
        outcome = sm()
        yasmin.YASMIN_LOG_INFO(outcome)
    except Exception as e:
        yasmin.YASMIN_LOG_WARN(e)

    if pub is not None:
        pub.shutdown()

    YasminNode.destroy_instance()

    # Shutdown ROS 2 if it's running
    if rclpy.ok():
        rclpy.shutdown()


if __name__ == "__main__":
    main()
